test/runtests.jl

```diff
@@ -228,310 +26,515 @@ catch
-let o = PyObject(Any[1,2]), c = o
-    o .+= Any[3,4]
-    @test collect(o) == [1,2,3,4]
-    @test o.o == c.o # updated in-place
-end
+    let x = 7
+        py"""
+        def myfun(x):
+            return x + $x
+        """
+        @test py"1 + 2" == 3
+        @test py"1 + $x" == 8
+        @test py"1 + $(x^2)" == 50
+        @test py"myfun"(10) == 17
+    end
 
-# more flexible bool conversions, matching Python "truth value testing"
-@test convert(Bool, PyObject(nothing)) === false
-@test convert(Bool, PyObject(0.0)) === false
-@test convert(Bool, PyObject(Any[])) === false
-@test convert(Bool, PyObject(17.3)) === true
-@test convert(Bool, PyObject(Any[0])) === true
-@test Bool(PyVector{PyObject}(PyObject([false]))[1]) === false
-
-# serialization
-let py_sum_obj = pybuiltin("sum")
-    b = IOBuffer()
-    serialize(b, py_sum_obj)
-    @test py_sum_obj == deserialize(seekstart(b))
-
-    b = IOBuffer()
-    serialize(b, PyNULL())
-    @test PyNULL() == deserialize(seekstart(b))
-end
+    # issue #352
+    let x = "1+1"
+        @test py"$x" == "1+1"
+        @test py"$$x" == py"$$(x)" == 2
+        @test py"7 - $$x - 7" == 0 # evaluates "7 - 1 + 1 - 7"
+        @test py"7 - ($$x) - 7" == -2 # evaluates "7 - (1 + 1) - 7"
+        @test py"1 + $$(x[1:2]) 3" == 5 # evals 1 + 1+ 3
+    end
 
-# issue #389
-@pydef mutable struct EmptyClass
-end
+    # Float16 support:
+    if PyCall.npy_initialized
+        @test roundtripeq(Float16[17 18 Inf -Inf -0.0 0.0])
+        @test isa(roundtrip(Float16[17]), Vector{Float16})
+    end
 
-# @pycall macro expands correctly
-_pycall = GlobalRef(PyCall,:pycall)
-@test macroexpand(@__MODULE__, :(@pycall foo(bar)::T)) == :($(_pycall)(foo, T, bar))
-@test macroexpand(@__MODULE__, :(@pycall foo(bar, args...)::T)) == :($(_pycall)(foo, T, bar, args...))
-@test macroexpand(@__MODULE__, :(@pycall foo(bar; kwargs...)::T)) == :($(_pycall)(foo, T, bar; kwargs...))
+    """
+    foobar doc
+    """
+    foobar(x) = x+1
+
+    # function attributes
+    let o = PyObject(foobar)
+        @test o[:__name__] == o[:func_name] == string(foobar)
+        @test o[:__doc__] == o[:func_doc] == "foobar doc\n"
+        @test o[:__module__] == o[:__defaults__] == o[:func_defaults] ==
+              o[:__closure__] == o[:func_closure] == nothing
+    end
 
+    # issue #345
+    let weakdict = pyimport("weakref")["WeakValueDictionary"]
+        # (use weakdict for the value, since Python supports
+        #  weak references to type objects)
+        @test convert(Dict{Int,PyObject}, weakdict(Dict(3=>weakdict))) == Dict(3=>weakdict)
+        @test get(weakdict(Dict(3=>weakdict)),3) == weakdict
+    end
 
-# basic @pywith functionality
-fname = tempname()
-try
-    @test begin
-        @pywith pybuiltin("open")(fname,"w") as f begin
-            f[:write]("test")
+    # Expose python docs to Julia doc system
+    py"""
+    def foo():
+        "foo docstring"
+        return 0
+    """
+    global foo = py"foo"
```

Comment:
Either `let foo=...` or just substitute `py"foo"` directly into the test since `foo` is used in only one place.

Reply:
```
using PyCall, BenchmarkTools, Base.Test
@testset "py docs" begin
    # Expose python docs to Julia doc system
    py"""
    def foo():
        "foo docstring"
        return 0
    """
    let foo = py"foo"
    # use 'content' since `Text` objects test equality by object identity
        @test @doc(foo).content == "foo docstring"
    end
end
```
Gives
```
py docs: Test Failed
  Expression: @doc(foo).content == "foo docstring"
   Evaluated: Any[Base.Markdown.Paragraph(Any["No documentation found."]), Base.Markdown.Paragraph(Any["Binding ", Base.Markdown.Code("", "foo"), " does not exist."])] == "foo docstring"
```
using `@doc(py"foo").content`
gives
```
Expression: @doc(@py_str("foo")).content == "foo docstring"
   Evaluated: Any[```
py".....python code....."

Evaluate the given Python code string in the main Python module.
...
```
Since
```
julia> @macroexpand @doc(py"foo").content
:((Base.Docs.doc)((Base.Docs.Binding)((current_module)(), Symbol("@py_str")), Union{Tuple{Any}}).content)
```

Seems like it needs to be defined on the module? Can I just change it to `global the_foo_function_in_python = py"foo"` ?

Reply:
Okay, I didn't realize `@doc` was so fragile.   Let's use `global foo354` then, since this was introduced in #354.

Reply:
I personally think `the_foo_function_in_python` has a better ring to it, possibly with an extra leading underscore, but ok, np 😄